frontend/catalyst/utils/wrapper.cpp

```diff
@@ -19,10 +190,14 @@ void wrap(py::object func, py::tuple py_args)
     void *value1_ptr = *reinterpret_cast<void **>(ctypes.attr("addressof")(value1).cast<size_t>());
 
     f_ptr(value0_ptr, value1_ptr);
+    returns = move_returns(value0_ptr, result_desc, transfer, numpy_arrays);
+
+    return returns;
 }
 
 PYBIND11_MODULE(wrapper, m)
 {
     m.doc() = "wrapper module";
     m.def("wrap", &wrap, "A wrapper function.");
+    _import_array();
```

Comment:
What function is this referencing?

Reply:
It's part of the numpy C-API. There is another function `import_array` (which in reality is a macro) but it is cumbersome to use from here since it returns a NULL pointer on failure. This convention of returning a NULL pointer on failure is very much in line with using the Python C-API but not with using pybind. I suppose a better approach here would be to keep `_import_array()` but check its return value (which in this case I think should be an integer) and raise an exception if it fails.

Reply:
Mmm... at compile time I saw an error saying that it was a macro that will return a `NULL` but will double check and see if this is a version issue.

Reply:
Here's the error I saw:

```
      /tmp/pip-build-env-q_q868w9/overlay/lib/python3.10/site-packages/numpy/core/include/numpy/__multiarray_api.h:1553:151: error: return-statement with a value, in function returning ‘void’ [-fpermissive]                                                        
       1553 | #define import_array() {if (_import_array() < 0) {PyErr_Print(); PyErr_SetString(PyExc_ImportError, "numpy.core.multiarray failed to import"); return NULL; } }                                                                                         
            |                                                                                                                                                       ^~~~                                                                                              
      frontend/catalyst/utils/wrapper.cpp:201:5: note: in expansion of macro ‘import_array’                                                                                                                                                                           
        201 |     import_array();                                                                                                                                                                                                                                     
            |     ^~~~~~~~~~~~   
```

This macro appears to be [generated at compile time / release time](https://github.com/numpy/numpy/blob/f45f692abac62265b760c0a249e8b417707c47d1/numpy/core/meson.build#L536-L537) so it is not available at the source.

But I found the definitions here:

```c

#define import_array() {if (_import_array() < 0) {PyErr_Print(); PyErr_SetString(PyExc_ImportError, "numpy.core.multiarray failed to import"); return NULL; } }

#define import_array1(ret) {if (_import_array() < 0) {PyErr_Print(); PyErr_SetString(PyExc_ImportError, "numpy.core.multiarray failed to import"); return ret; } }

#define import_array2(msg, ret) {if (_import_array() < 0) {PyErr_Print(); PyErr_SetString(PyExc_ImportError, msg); return ret; } }

```

I do think that using `_import_array()` directly and not these macros is not a bad idea. Otherwise, if you insist then I can instead create a function which calls the macro `import_array1` and check the return code there.

Reply:
That's interesting, it even says that the function returns void in the error message: `error: return-statement with a value, in function returning ‘void’ [-fpermissive]`. Is that a bug in numpy, or is this only illegal in C++?
Looks like GCC will allow it in C while issuing a warning, still seems like a fishy thing to do 🤔

Reply:
> That's interesting, it even says that the function returns void in the error message: error: return-statement with a value, in function returning ‘void’ [-fpermissive]

The function that returns void in this case is the function who is calling `import_array`.

Reply:
That would explain it then. I guess they expect the function to be called from `int main()`.
So we can either wrap the call in a function that returns a number, or just use the expanded macro but without the `return NULL;` part.

Reply:
Not from `int main()`. It is expected to be called from the Python C-API as opposed to pybind. A function in a module defined using the Python C-API is expected to return a `Py_Object*` always. (Just like a function in python always returns `None` even in the absence of a return statement.) However, the Python C-API expects that NULL will be returned in case of error an exception is set using one of the error functions.